add karmada cert message doc (#824)
karmada-io/website: docs/administrator/security/cert-framework.md

---
title: Certificate Framework
---

# Karmada Certificate Framework

This document provides a comprehensive description of the certificate framework in the Karmada system, including the organizational structure, purpose, and configuration recommendations for all certificate resources. The Karmada certificate framework defines the certificate system required for secure communication between components, clearly articulating how each component uses certificates for authentication and communication encryption. This document will help administrators understand Karmada's certificate architecture, correctly configure and manage the certificates required by components, and ensure the security of the entire system.

> Note: Currently, when installing Karmada using the community-maintained `hack/deploy-karmada.sh` script, certificates are generated following the framework described in this document. Other [installation methods](../../installation/installation.md)—including `karmadactl init`, `karmada-operator`, and Helm—will align with this certificate standard in future releases. For more details about this change, see [this issue](https://github.com/karmada-io/karmada/issues/6091).

Author: Okay, I’ve fixed these minor errors in both the English and Chinese versions of the documents. How do they look now? Could you help check?

> Currently, this document introduces the certificate information related to secrets in the cluster deployed using the community-maintained `hack/deploy-karmada.sh` script installation tool. Since there are currently four deployment methods in the Karmada system (deployment via `karmadactl init`, deployment via `karmada-operator`, deployment via `helm`, and deployment via the `hack/deploy-karmada.sh` script installation tool), in Karmada versions `v1.13` and earlier, components in the Karmada cluster shared the same certificate content. This will be a preliminary change to standardize and implement the new Karmada certificate standard (information about this change can be found in the [related issue](https://github.com/karmada-io/karmada/issues/6091)).

Reviewer: How about:

```suggestion
Note: Currently, certificates are generated according to the certificate framework described in this document when installing Karmada via the community-maintained `hack/deploy-karmada.sh` script. Other [installation methods](../../installation/installation.md) will be synchronized with this document in the future.
```

Author: OK, I have updated this part of the description as suggested. Does the content look appropriate now?

Here is an example of the Karmada root certificate:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            5a:f6:2d:21:a5:31:92:99:32:07:00:dd:8f:37:0a:c0:ae:02:2f:c1
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=karmada
        Validity
            Not Before: %b %d %H:%M:%S %Y GMT
            Not After : %b %d %H:%M:%S %Y GMT
        Subject: CN=karmada
        Subject Public Key Info:
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
```

## Karmada API Server

### Server Certificate

The Server Certificate for Karmada API Server is used to encrypt and authenticate communication between the `karmada-apiserver` and clients. This certificate ensures that the `karmada-apiserver` can securely provide its service interface, while allowing clients to verify the server's authenticity, preventing man-in-the-middle attacks.

> **Command Line Flags**
> Passed via the `--tls-cert-file` flag

Reviewer 2: ```suggestion
**Command Line Flags**

Passed via the `--tls-cert-file` flag.
```

**Recommended CN and SAN Values**

```
Subject: CN=system:karmada:karmada-apiserver
Subject Public Key Info:
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Client Authentication, TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name:
                DNS:karmada-apiserver.karmada-system.svc.cluster.local,
                DNS:karmada-apiserver.karmada-system.svc, DNS:localhost,
                IP Address:127.0.0.1, IP Address:${apiserver_service_ip_address}
```

### ETCD Client Certificate

The ETCD Client Certificate for Karmada API Server is used for identity authentication when the `karmada-apiserver` accesses the ETCD data store. It ensures that only authorized API Server instances can read and write data in ETCD.

> **Command Line Flags**
> Passed via the `--etcd-certfile` flag

Reviewer 2: ```suggestion
**Command Line Flags**

Passed via the `--etcd-certfile` flag.
```

**Command Line Flags**
Passed via the `--proxy-client-cert-file` flag

**Recommended CN and SAN Values**

```
Subject: CN=front-proxy-client
Subject Public Key Info:
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Client Authentication, TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
```

## Karmada Aggregated API Server

### Server Certificate

The Server Certificate for Karmada Aggregated API Server is used for Karmada's API aggregation layer, allowing API extensions to be dynamically added to the core API Server. This certificate ensures the identity authentication of the Aggregated API Server and secures communication with other components. Through the `karmada-aggregated-apiserver`, Karmada can securely extend its API functionality.

**Command Line Flags**

> Passed via the `--tls-cert-file` flag

Reviewer 2: Same as above

**Recommended CN and SAN Values**

```
Subject: CN=system:karmada:karmada-aggregated-apiserver
Subject Public Key Info:
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Client Authentication, TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name:
                DNS:karmada-aggregated-apiserver.karmada-system.svc.cluster.local, 
                DNS:karmada-aggregated-apiserver.karmada-system.svc, DNS:localhost, 
                IP Address:127.0.0.1
```

### Client Certificate

The Client Certificate for Karmada Aggregated API Server is used for identity authentication when the `karmada-aggregated-apiserver` component communicates with the `karmada-apiserver` as a client. It allows the Karmada Aggregated API Server component to securely access API services and perform necessary operations. This certificate ensures that Karmada Aggregated API Server requests are correctly identified and authorized.

**Command Line Flags**

> Passed via the `--kubeconfig` flag containing the certificate

Reviewer 2: Same as above

**Recommended CN and SAN Values**

```
Subject: O=system:masters, CN=system:karmada:karmada-aggregated-apiserver
Subject Public Key Info:
        X509v3 extensions:
            X509v3 Subject Alternative Name:
```

### ETCD Client Certificate

The ETCD Client Certificate for Karmada Aggregated API Server is used for identity authentication when the `karmada-aggregated-apiserver` accesses the ETCD data store. It ensures that the `karmada-aggregated-apiserver` can securely store and retrieve the state information of its extended API resources while maintaining appropriate access control.

**Command Line Flags**

> Passed via the `--etcd-certfile` flag

Reviewer 2: Same as above

**Recommended CN and SAN Values**

```
Subject: O=system:masters, CN=system:karmada:karmada-aggregated-apiserver-etcd-client
Subject Public Key Info:
        X509v3 extensions:
            X509v3 Subject Alternative Name:
```

## Karmada Webhook

### Server Certificate

The Server Certificate for Karmada Webhook is used to encrypt and authenticate communication between the `karmada-webhook` and the `karmada-apiserver`. The `karmada-webhook` is responsible for resource validation and admission control in Karmada, ensuring that all operations on resources comply with predefined rules.

**Command Line Flags**

> Passed via the `--cert-dir` flag, automatically read from the directory

Reviewer 2: Same as above
